Make transform a method

jobrachem opened the pull request:
This PR introduces a `Var.transform` method. Some notes:

### Replacing `GraphBuilder.transform`

We currently have `GraphBuilder.transform`. The new method is in fact intended as a replacement for `GraphBuilder.transform`. Having `Var.transform` as a method on `Var` has the advantage that it is easier to find for users. Since the transformation is "doing something to a Var" and it does not actually require any functionality within the GraphBuilder, living as a method on a Var is a natural development for the transform method.

### Behavior change

The method behaves similar to `GraphBuilder.transform`, with a few notable differences:

1. It will *not* use the default event space bijector from tensorflow by default. This change is made to encourage users to either select their desired bijector manually, which is often sensible, or to request an automatic bijector manually. In both cases, users are more aware of what they are doing.
2. The new method accepts bijector *instances* in addition to bijector *classes*. In fact, passing an instance is the preferred way of passing a bijector. Passing a bijector class is only supported if you actually defined `*bijector_args` or `*bijector_kwargs` to be passed to the bijector. This simplifies the code for the default case. More importantly, this fixes the graph representation after transformation, see below.


Using a bijector instance:

<img width="950" alt="image" src="https://github.com/liesel-devs/liesel/assets/37882800/802ffaf4-fc97-4033-99e0-20dd9619fb42">


Compare to using the default event space bijector. Note that in this case, the same bijector is being used, but there are spurious edges from the nodes "v0" and "v1", the prior parameters, to the original variable "tau".

<img width="940" alt="image" src="https://github.com/liesel-devs/liesel/assets/37882800/7a021ef4-4613-4a6e-aa58-bbc8ce11366b">

### Deprecation and updated documentation

1. I marked `GraphBuilder.transform` as deprecated and included directions towards the new method in its documentation.
2. I updated the usage of the transformation method in the tutorials `01a-transform.md` and `04-mcycle.md`. As it turned out, usage in these tutorials was outdated anyway.

### Notebook for testing

You can play around with the method in this notebook:

[050-transform.ipynb.zip](https://github.com/liesel-devs/liesel/files/13779942/050-transform.ipynb.zip)


### Related issues

- This PR closes #173 
- This PR closes #93
- This PR improves upon the behavior described in #89, because the problem described in #89 does not occur if a bijector instance is passed.

jobrachem commented:
A log message on the default bijector being used would be nice.

jobrachem commented:
@jobrachem will transfer the core benefits of the `Var.transform` method to the `GraphBuilder.transform`: 1) accepting a bijector instance, 2) logging the default event space bijector

jobrachem commented:
@jobrachem will continue as follows:

1. Implement Var.transform
2. Make API between Var.transform and GraphBuilder.transform consistent
3. Add a note to the docs of GraphBuilder.transform, referring to Var.transform

jobrachem commented:
I now harmonized the API between `Var.transform` and `GraphBuilder.transform`. Namely:

1. `Var.transform` now has the same default behavior as `GraphBuilder.transform`, i.e. it tries to use the default event space bijector from tensorflow probability.
2. `GraphBuilder.transform` now also accepts a bijector instance.

There is one small difference in how the methods handle it, if the user passes a bijector *class*: `Var.transform` will raise a RuntimeError, if you pass a bijector class, but do not actually use any arguments for the bijector. `GraphBuilder.transform` will emit a `UserWarning` instead, in order to avoid breaking existing code.

jobrachem commented:
@wiep gentle reminder :)

wiep commented on liesel/model/model.py:
```diff
@@ -799,6 +828,17 @@ def transform(
+            bijector, type
+        )
+
+        if isinstance(bijector, type) and not args and not kwargs:
```
might use

`if isinstance(bijector, type) and not (args or kwargs):`

for better readability.

wiep commented on liesel/model/model.py:
```diff
@@ -758,12 +773,26 @@ def transform(
         if var.dist_node is None:
             raise RuntimeError(f"{repr(var)} has no distribution")
 
+        if bijector is not None and not isinstance(bijector, type) and (args or kwargs):
```
can we test here for `isintance(bijector, BijectorType)` instead of `bijector is not None and not isinstance(bijector, type)`?

would make it better comprehensible

jobrachem replied:
I don't know about a `BijectorType` object, but I clarified the check by introducing

```python
def is_bijector_class(obj):
    return isinstance(obj, type) and issubclass(obj, jb.Bijector)
```

and using

```python
if not is_bijector_class(bijector) and (args or kwargs):
```

now :)

wiep commented on liesel/model/nodes.py:
```diff
@@ -1100,6 +1103,159 @@ def all_input_vars(self) -> tuple[Var, ...]:
+
+        # avoid infinite recursion
+        self.auto_transform = False
+
```
could handle the `bijector is None` case above line 1233 and remove duplicated code:
```
if bijector is None:
            dist_inst = self.dist_node.init_dist()
            bijector = dist_inst.experimental_default_event_space_bijector
            
if isinstance(bijector, type):
            tvar = _transform_var_with_bijector_class(
                self, bijector, *bijector_args, **bijector_kwargs
            )
else: #bijector is bijector instance
   ...
```

jobrachem replied:
See my comment below, this  suggestion alerted me to a problem in my implementation. Thanks!

wiep commented on liesel/model/nodes.py:
```diff
@@ -1100,6 +1103,159 @@ def all_input_vars(self) -> tuple[Var, ...]:
+            tvar = _transform_var_with_bijector_class(
+                self, bijector, *bijector_args, **bijector_kwargs
+            )
+        else:
```
Should we test for the correct type by using:

```
elif isinstance(bjector, BijectorType):
    ...
else:
    raise ValueError("bijector must be None | instance of bijectorType | BijectorType")
```

jobrachem replied:
I implemented that as follows:

```python
else:
    raise TypeError(
        f"Argument {bijector=} is of invalid type {type(bijector)}."
    )
```

wiep commented on liesel/model/nodes.py:
```diff
@@ -1100,6 +1103,159 @@ def all_input_vars(self) -> tuple[Var, ...]:
+        if self.dist_node is None:  # type: ignore
+            raise RuntimeError(f"{repr(self)} has no distribution")
+
+        if isinstance(bijector, type) and not bijector_args and not bijector_kwargs:
```
similar to above
`and not (bijector_args or bijector_kwargs)`

wiep commented on liesel/model/nodes.py:
```diff
@@ -1386,6 +1542,92 @@ def __repr__(self) -> str:
+    return transformed_var
+
+
+def _transform_var_with_bijector_class(
```
to confirm my understanding: we cannot just instanciate  the bijector here and then use `_transform_var_with_bijector_instance` because 

1) the inputs may not be available at this time
2) the bijector may change when the inputs to the bijector change

jobrachem replied:
Yes, and I think point two is the critical one: Even of the input values were available at the time that `.transform` is called, we do not want to use their concrete value here. Their values may change, and as a result, the bijector's behavior may change. So when a bijector class is used, we have to set up a wrapping system that initializes the bijector anew during every update with the current input values.

wiep reviewed (commented):
Thanks for the work.

I made a few suggestions to make the if statement better comprehensible. I think I would handle the bijector is None case such that I would cast it into a bijector is instance of BijectorType case. Than you need to handle only two cases. See annotations below.

However, both is a bit of a style question. Should you prefer the current version. I'm also ok with that.


Also, please feel free to revert the commit changing doc strings if you do not agree (dfb5217e381b58878c0c99fc901bba0a765ea0a1).

jobrachem commented:
Hey @wiep, thanks a lot for your suggestions! I implemented them as closely as possible. Your change to the docstring is fine with me, too 👍 

Your comment about removing the code duplication also alerted me to two issues:

1. `Var.transform` lacked a test for whether the default event space bijector exists. I added that test.
2. There was a subtle error in my implementation of using the default event space bijector. In the case of using the default bijector, we want to account for the possibility that the default bijector can change depending on the distribution's arguments. So the default bijector has to be obtained and initialized anew for every `.update()` call. This was not the case in my previous implementation. Not it is, so this problem is fixed.
